flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3317,6 +3317,39 @@ class ArrayExprLowering {
     return genarr(asScalarRef(sym));
   }
 
+  /// Base the indices according to the origin of an entity. This is needed for
+  /// operations like fir.array_coor that do not expect zero based indices.
+  llvm::SmallVector<mlir::Value>
+  applyShift(mlir::Value shape, mlir::Value slice,
```

Comment:
This should actually share the code in `originateIndices()` in ArrayValueCopy.cpp, I think.

Reply:
Good point, I am not sure how best to share this given one is using a FirOpBuilder, and the other a mlir::PatternRewritter, so I shared the code as template code in `Optimizer/Transforms/Factory.h`. Does that sounds OK to you ?